client/asset/zec/zec.go

```diff
@@ -224,7 +238,229 @@ func NewWallet(cfg *asset.WalletConfig, logger dex.Logger, net dex.Network) (ass
+			}
+			return fmt.Errorf("error creating account %d: %w", n, err)
+		}
+		if acctNumber != n {
+			return fmt.Errorf("no account %d found and newly created account has unexpected account number %d", n, acctNumber)
+		}
+		return nil
+	}
+	if err := createAccount(0); err != nil {
+		return err
+	}
+	if err := createAccount(1); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ShieldedBalances list the address-balance pairs for the wallet's shielded
+// account.
+func (w *zecWallet) ShieldedBalances() ([]*asset.ShieldedBalance, error) {
+	// z_listaccounts to get account 1 addresses
+	// DRAFT NOTE: It sucks that we need to list all accounts here. The zeroth
+	// account is our transparent addresses, and they'll all be included in the
+	// result. Should probably open a PR at zcash/zcash to add the ability to
+	// list a single account.
+	accts, err := zListAccounts(w)
+	if err != nil {
+		return nil, err
+	}
+
+	res := make([]*asset.ShieldedBalance, 0)
+	for _, acct := range accts {
+		if acct.Number != shieldedAcctNumber {
+			continue
+		}
+		for _, a := range acct.Addresses {
+			recv, err := zGetUnifiedReceivers(w, a.UnifiedAddr)
+			if err != nil {
+				return nil, err
+			}
+			bal, err := zGetBalance(w, recv.Orchard)
+			if err != nil {
+				return nil, fmt.Errorf("error getting balance for Orchard address %s: %w", recv.Orchard, err)
+			}
+			res = append(res, &asset.ShieldedBalance{
+				Address: recv.Orchard, // Same as a.UnifiedAddress in account 1
+				Balance: bal,
+			})
+		}
+	}
+	return res, nil
+}
+
+// Balance adds a sum of shielded pool balances to the transparent balance info.
+func (w *zecWallet) Balance() (*asset.Balance, error) {
+	bal, err := w.ExchangeWalletFullNode.Balance()
+	if err != nil {
+		return nil, err
+	}
+
+	bals, err := w.ShieldedBalances()
+	if err != nil {
+		return nil, err
+	}
+
+	var shielded uint64
+	for _, bal := range bals {
+		shielded += bal.Balance
+	}
+
+	if bal.Other == nil {
+		bal.Other = map[string]uint64{}
+	}
+
+	bal.Other["shielded"] = shielded
+	return bal, nil
+
+}
+
+// NewShieldedAddress creates a new shielded address. Unlike regular utxo
+// assets, you do not create a shielded address for every transaction. A
+// shielded address is to be reused, so creating a new address is more like
+// creating a new account.
+func (w *zecWallet) NewShieldedAddress() (string, error) {
+	// An orchard address is the same as a unified address with only an orchard
+	// receiver.
+	addrRes, err := zGetAddressForAccount(w, shieldedAcctNumber, []string{orchardAddressType})
+	if err != nil {
+		return "", err
+	}
+	return addrRes.Address, nil
+}
+
+// ShieldFunds moves funds from the transparent account to the specified
+// shielded address.
+func (w *zecWallet) ShieldFunds(ctx context.Context, shieldedAddr string, transparentVal uint64) (dex.Bytes, error) {
```

Comment:
Since we're only doing one account, why should `core` have to deal with the various shielded addresses? It seems like this should just be managed by the wallet and there is no need for a `shieldedAddr` argument.

Reply:
One account can create multiple addresses, but shielded addresses are not single-use like trasnparent addresses, so a shielded address itself acts like an account. Users will be able to create new shielded addresses.

Reply:
I assumed you would be able to use the funds from multiple addresses in the same account in the same transaction, since they are all controlled by the same private key. What's the point of having the same public key control the funds of multiple addresses if you can't send them in the same transaction?

Reply:
I was talking to ChatGPT and it told me that all of the addresses in an account are derived from the same private key, so I assumed you could send funds from multiple addresses in the same tx. Gotta stop talking to ChatGPT lol.

Reply:
So I was wrong. Balances and spending are associated with the account, not the particular address. When using `z_sendmany`, you can provide any address from the account as the `fromaddress` argument. Same thing for the deprecated `z_getbalance` method. Doesn't matter which address you specify, as long as it's from the right account. Had me all confused. Fixed now.

Reply:
So ChatGPT was right.

Reply:
:robot: